feat(docs): upgrade guide for v0.6 (#991)
hashicorp/terraform-cdk: docs/upgrade-guide/upgrading-to-0.6.md

While using the v0.6 `cdktf-cli` to deploy stacks that are using an older version of `cdktf` and `constructs` should still work, we recommend upgrading the `cdktf` and `constructs` packages.

## Pre-built providers

All upcoming releases of the [pre-built providers](https://cdk.tf/prebuilt-providers) will depend on `constructs` v10. So we recommend to upgrade your project.
If you cannot upgrade the `cdktf` version, you can switch to [generating provider bindings](https://github.com/hashicorp/terraform-cdk/blob/main/docs/working-with-cdk-for-terraform/using-providers.md#importing-providers-and-modules) via `cdktf get`.

## Upgrading a TypeScript / JavaScript CDKTF project

To upgrade a TypeScript or JavaScript CDKTF project, adjust the versions in the `package.json` file of your project.

```jsonc
{
  // ...
  "dependencies": {
    "cdktf": "^0.6.0",

>     "constructs": "^10.0.5"

Reviewer: Do we warn users if they run 0.6 using an older version of the constructs package?

Author: The peer dependency is set to v10, so e.g. for TS/JS there will be a warning from npm / yarn. For the other languages I don't know this but it could have similar effects.
However, we could add a check to the `cdktf` package which could then fail and print an useful error message 👍
Going to leave this discussion open until I added such a check.

To upgrade a Python CDKTF project using `pip3`, adjust the versions in the `requirements.txt` of your project.

```
cdktf~=0.6.0
```

Afterwards run `pip3 install -r requirements.txt` to update the depedencies and `cdktf get` to re-generate the provider bindings if you don't only use pre-built providers.

## Upgrading a Java CDKTF project

To upgrade a Java CDKTF project, adjust the versions in the `pom.xml` file of your project.

```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <dependencies>
    <dependency>
      <groupId>com.hashicorp</groupId>
      <artifactId>cdktf</artifactId>
      <version>0.6.0</version>
    </dependency>
    <dependency>
      <groupId>software.constructs</groupId>
      <artifactId>constructs</artifactId>
      <version>10.0.5</version>
    </dependency>
  </dependencies>
  <!-- ... -->
</project>
```

Afterwards run `mvn install` to update the depedencies and `cdktf get` to re-generate the provider bindings if you don't only use pre-built providers.

## Upgrading a CSharp CDKTF project

> To upgrade a Java CDKTF project, adjust the versions in the `*.csproj` file of your project.

Reviewer 2: ```suggestion
To upgrade a C# CDKTF project, adjust the versions in the `*.csproj` file of your project.
```

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <!-- ... -->
  <ItemGroup>
    <PackageReference Include="HashiCorp.Cdktf" Version="0.6.0" />
  </ItemGroup>
</Project>
```

Afterwards run `dotnet restore` to update the depedencies and `cdktf get` to re-generate the provider bindings if you don't only use pre-built providers.

## Upgrading a Golang CDKTF project

> To upgrade a Java CDKTF project, adjust the versions in the `go.mod` file of your project.

Reviewer 2: ```suggestion
To upgrade a Golang CDKTF project, adjust the versions in the `go.mod` file of your project.
```

import { Aspects } from "cdktf"; // new
Aspects.of(construct).add(aspect);

// instead of
construct.node.applyAspect(aspect); // old
```

### `construct.node.addInfo()` / `addWarning()` / `addError()` moved to `Annotations`

The aforementioned methods are now available via the `Annotations` API:

```typescript
// use
import { Annotations } from "cdktf"; // new
Annotations.of(construct).addInfo("hello");

// instead of
construct.node.addInfo("hello"); // old
```

Annotations will also be printed by the `cdktf-cli` after your stack was synthesized. One ore more `error` annotations will make the `synth` fail.

### `onPrepare()` and `prepare()` have been removed

```typescript
// use
import { Aspects } from "cdktf";
class MyConstruct extends Construct {
  constructor(scope: IConstruct, id: string) {
    super(scope, id);

    Aspects.of(this).add({
      visit: (node: IConstruct) => {
        // visit will also be called on any childs of this construct
        if (node === this) console.log("some work");
      },
    });
  }
}

// instead of
class MyConstruct extends Construct {
  onPrepare() {
    console.log("some work");
  }
}
```

### `onSynthesize` has been removed

> Should you find yourself needing support for this, please file a [new issue](https://cdk.tf/feature) on this repo and describe your use-case.

Reviewer 3: ```suggestion
Should you need support for this, please file a [new issue](https://cdk.tf/feature) on this repo and describe your use-case.
```
